tests/core/test_EIP712.py

```diff
@@ -0,0 +1,437 @@
+        "d872b041d703d1b5aadf3154a261abdd9086fc627b61efca26ae5702701d05cd2305f7c52a2fc8"
+    )
+    assert encodeData(primary_type, types, message).hex() == expected_hex_value
+
+
+def test_hashStruct_main_message(structured_valid_data_json_string):
+    expected_hex_value = "c52c0ee5d84264471806290a3f2c4cecfc5490626bf912d01f240d7a274b371e"
+    assert hashStruct(structured_valid_data_json_string).hex() == expected_hex_value
+
+
+def test_hashStruct_domain(structured_valid_data_json_string):
+    expected_hex_value = "f2cee375fa42b42143804025fc449deafd50cc031ca257e0b194a650a912090f"
+    assert (
+        hashStruct(structured_valid_data_json_string, for_domain=True).hex() == expected_hex_value
+    )
+
+
+def test_hashed_structured_data(signature_kwargs):
+    hashed_structured_msg = defunct_hash_message(
+        **signature_kwargs,
+        signature_version=b'\x01',
+    )
+    expected_hash_value_hex = "0xbe609aee343fb3c4b28e1df9e632fca64fcfaede20f02e86244efddf30957bd2"
+    assert hashed_structured_msg.hex() == expected_hash_value_hex
+
+
+def test_signature_verification(signature_kwargs):
+    account = Account.create()
+    hashed_structured_msg = defunct_hash_message(
+        **signature_kwargs,
+        signature_version=b'\x01',
+    )
+    signed = Account.signHash(hashed_structured_msg, account.privateKey)
+    new_addr = Account.recoverHash(hashed_structured_msg, signature=signed.signature)
+    assert new_addr == account.address
+
+
+def test_signature_variables(signature_kwargs):
+    # Check that the signature of typed message is the same as that
+    # mentioned in the EIP. The link is as follows
+    # https://github.com/ethereum/EIPs/blob/master/assets/eip-712/Example.js
+    hashed_structured_msg = defunct_hash_message(
+        **signature_kwargs,
+        signature_version=b'\x01',
+    )
+    privateKey = keccak(text="cow")
+    acc = Account.privateKeyToAccount(privateKey)
+    assert HexBytes(acc.address) == HexBytes("0xcd2a3d9f938e13cd947ec05abc7fe734df8dd826")
+    sig = Account.signHash(hashed_structured_msg, privateKey)
+    assert sig.v == 28
+    assert hex(sig.r) == "0x4355c47d63924e8a72e509b65029052eb6c299d53a04e167c5775fd466751c9d"
+    assert hex(sig.s) == "0x7299936d304c153f6443dfa05f40ff007d72911b6f72307f996231605b91562"
+
+
+@pytest.mark.parametrize(
+    'type, valid',
+    (
+        ("unint bytes32", False),
+        ("hello\\[]", False),
+        ("byte[]uint", False),
+        ("byte[7[]uint][]", False),
+        ("Person[0]", False),
+
+        ("bytes32", True),
+        ("Foo[]", True),
+        ("bytes1", True),
+        ("bytes32[][][]", True),
+        ("byte[9]", True),
+        ("Person[1]", True),
+    )
+)
+def test_type_regex(type, valid):
+    if valid:
+        assert re.match(TYPE_REGEX, type) is not None
+    else:
+        assert re.match(TYPE_REGEX, type) is None
+
+
+def test_structured_data_invalid_identifier_filtered_by_regex():
+    invalid_structured_data_string = '''{
```

Comment:
I don't really have any good ideas about how to do this right at the moment, but it seems like there should be a way to be able to merge these invalid data structures with `structured_valid_data_json_string`. So instead of re-writing the whole string out, we could do something like: 
```
structured_valid_data_json_string.merge(
  {"Person": [ "name": "hello wallet", "type": "address" }] }
)
```

Reply:
I'm in search of such a method.

Reply:
@pipermerriam I was wondering how to go about this. How about we maintain a lot of `JSON` files in the `tests` directory and load them as `fixtures`, instead of writing those JSON strings in the test file? (This is similar to what we are doing in `py-evm` and `trinity`)

Reply:
I'm :+1: on using individual JSON files in an appropriately placed `./fixtures/` directory (probably alongside these tests).